dynamic-check/src/objects/frame.rs

```diff
@@ -0,0 +1,1198 @@
+        let circuit_function_type = self.parse_circuit_function_type(expression, identifier, span)?;
+
+        // Check that the function is non-static.
+        if circuit_function_type.attributes.contains(&Attribute::Static) {
+            return Err(FrameError::invalid_static_access(identifier));
+        }
+
+        // Return the function type.
+        Ok(circuit_function_type.function.to_owned())
+    }
+
+    ///
+    /// Returns a `FunctionType` given a circuit expression and static function identifier.
+    ///
+    fn parse_static_circuit_function(
+        &mut self,
+        expression: &Expression,
+        identifier: &Identifier,
+        span: &Span,
+    ) -> Result<FunctionType, FrameError> {
+        // Find circuit function type.
+        let circuit_function_type = self.parse_circuit_function_type(expression, identifier, span)?;
+
+        // Check that the function is static.
+        if !circuit_function_type.attributes.contains(&Attribute::Static) {
+            return Err(FrameError::invalid_member_access(identifier));
+        }
+
+        Ok(circuit_function_type.function.to_owned())
+    }
+
+    ///
+    /// Returns the type returned by calling the function.
+    ///
+    /// Does not attempt to evaluate the function call. We are just checking types at this step.
+    ///
+    fn parse_function_call(
+        &mut self,
+        expression: &Expression,
+        inputs: &Vec<Expression>,
+        span: &Span,
+    ) -> Result<Type, FrameError> {
+        // Parse the function name.
+        let function_type = self.parse_function_name(expression, span)?;
+
+        // Check the length of arguments
+        if function_type.inputs.len() != inputs.len() {
+            return Err(FrameError::num_inputs(function_type.inputs.len(), inputs.len(), span));
+        }
+
+        // Assert function inputs are correct types.
+        for (expected_input, actual_input) in function_type.inputs.iter().zip(inputs) {
+            // Parse expected input type.
+            let expected_type = expected_input.type_();
+
+            // Parse actual input type.
+            let actual_type = self.parse_expression(actual_input)?;
+
+            // Assert expected input type == actual input type.
+            self.assert_equal(expected_type, actual_type, span);
+        }
+
+        // Return the function output type.
+        Ok(function_type.output.type_.clone())
+    }
+
+    ///
+    /// Returns the type returned by calling the core function.
+    ///
+    fn parse_core_function_call(
+        &mut self,
+        _name: &String,
+        _arguments: &Vec<Expression>,
+        _span: &Span,
+    ) -> Result<Type, FrameError> {
+        unimplemented!("type checks for core function calls not implemented")
+    }
+
+    ///
+    /// Returns `Ok` if all `TypeAssertions` can be solved successfully.
+    ///
+    pub(crate) fn check(self) -> Result<(), FrameError> {
+        let mut unsolved = self.type_assertions.clone();
```

Comment:
since `self` is consumed here, this clone shouldn't be necessary; if the borrow checker complains, `self` might need to be destructured via pattern-matching